### ISSUE TITLE
OSGi: soft proxy creation lacks visibilty to internal types

### ISSUE THREAD

**Issue by bjhargrave** [joel-costigliola/assertj-core#1964]
In the [osgi-test](https://github.com/osgi/osgi-test) project, we are developing custom AssertJ assertions and providing soft assertion support for them. [Integration testing for the latest such custom assertion](https://github.com/osgi/osgi-test/pull/176), show a class loading issue preventing creating the soft assertion proxy. The issue seems to be at  

https://github.com/joel-costigliola/assertj-core/blob/0cadf3e19ff126fc96f2055d9923b6fb327a02be/src/main/java/org/assertj/core/api/SoftProxies.java#L198-L200

where it uses the class loader of the assertClass but the OSGi bundle holding the assertClass does not have visibility to the internal packages of AssertJ which also need to be visible to the ByteBuddy generated class. So when the integration test runs, we get the following error:

https://github.com/osgi/osgi-test/pull/176/checks?check_run_id=972751583#step:4:727
```
[INFO] --- bnd-testing-maven-plugin:5.1.2:testing (testing) @ org.osgi.test.junit5 ---
[INFO] Matching glob *
[INFO] Bnd inferred -runee: JavaSE-1.8
TEST org.osgi.test.junit5.dictionary.DictionaryAssertTest#testDictionarySoftAssertion() <<< ERROR: org/assertj/core/internal/TypeComparators
java.lang.NoClassDefFoundError: org/assertj/core/internal/TypeComparators
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethod(Class.java:2128)
	at org.osgi.test.assertj.dictionary.ProxyableDictionaryAssert$ByteBuddy$QRu2iZHZ.<clinit>(Unknown Source)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.assertj.core.api.SoftProxies.createSoftAssertionProxy(SoftProxies.java:118)
	at org.assertj.core.api.AbstractSoftAssertions.proxy(AbstractSoftAssertions.java:43)
	at org.osgi.test.assertj.dictionary.DictionarySoftAssertionsProvider.assertThat(DictionarySoftAssertionsProvider.java:32)
	at org.osgi.test.junit5.dictionary.DictionaryAssertTest.testDictionarySoftAssertion(DictionaryAssertTest.java:22)
```

The ByteBuddy generated class needs to use a class loader which can load classes from the class loader of the assertion class (`assertClass.getClassLoader()`) and also from the class loader of AssertJ so that internal types can be found. So I think the supplied class loader here needs to be a composite class loader over both `assertClass.getClassLoader()` and `getClass().getClassLoader()`.

**Comment by kriegfrj** [comment]
I think this is also another :+1: for #1888.

**Comment by joel-costigliola** [comment]
@bjhargrave thanks for the detailed issue, is that something you can work on? 
we don't really have good testing when it comes to OSGi as @kriegfrj pointed out with a reference to #1888

**Comment by bjhargrave** [comment]
OK, so my experiments with trying to use a composite class loader have failed. The main issues is that AssertJ uses a ByteBuddy class loading strategy which defines the proxy class in the class loader which defines the assertClass. For OSGi, this is the bundle's class loader. But that class loader does not have visibility to AssertJ internal packages in the assertj-core bundle and so we get the exception above.

Trying to use a composite class loader does not help since such a class loader is not a defining class loader. It is merely delegating. So once the class is loaded, the class' defining class loader is used to do further class loading such as super classes. And the composite class loader is out of the loop. The defining class loader for the generated proxy class requires visibility to the internal packages at least as shown here in the present case where the custom assert class is a subclass of AbstractMapAssert. We have two other custom asserts in osgi-test which do not exhibit this issue as they subclass different AssertJ types. So this may have something to do with the implementation of AbstractMapAssert and the way AssertJ configures the proxy generation for ByteBuddy.

Trying to use a ByteBuddy class loading strategy which uses a ByteBuddy created class loader to host the proxy class ("WRAPPER") causes numerous other problems as this breaks package-private access (which cannot "reach" across class loaders).

I think the only way out of this problem is for the assertj-core bundle to Export-Package _all_ packages including the "internal" packages. While this does not comport with "best practices", it seems the only way for a dynamically generated proxy to have visibility to all of the packages and package-private code in the bundle defining the custom assert class and all of the necessary internal types in assertj-core (since the bundle defining the custom assert class can import, and thus have visibility, to any necessary types in assertj-core). That is, we must provide the bundle's class loader visibility to all assertj-core types.

**Comment by kriegfrj** [comment]
I just started to see if I could have a play with this, because I thought that @bjhargrave's composite classloader idea should work.

I happened to notice this in the Maven log output for `bnd-maven-plugin`:

[INFO] --- bnd-maven-plugin:5.1.2:bnd-process (bundle-after-unpack-shaded) @ assertj-core ---
[WARNING] C:\workspace\assertj-core\dependency-reduced-pom.xml [0:0]: Export org.assertj.core.api,  has 2,  private references [org.assertj.core.internal, org.assertj.core.internal.bytebuddy.implementation.bind.annotation]
[WARNING] C:\workspace\assertj-core\dependency-reduced-pom.xml [0:0]: Export org.assertj.core.api.recursive.comparison,  has 1,  private references [org.assertj.core.internal]
[WARNING] C:\workspace\assertj-core\dependency-reduced-pom.xml [0:0]: Export org.assertj.core.error,  has 1,  private references [org.assertj.core.internal]
[WARNING] C:\workspace\assertj-core\dependency-reduced-pom.xml [0:0]: Export org.assertj.core.presentation,  has 1,  private references [org.assertj.core.internal]

These warnings are genuine problems that need to be fixed, and I think that they may be the root cause of the issues that BJ has identified.

The warning means (using the first one an example) that there is a class in public package `org.assertj.core.api` which contains public classes that include references to classes in those other private packages as part of their class definition (eg, it inherits from one of them, or contains one in a method signature).

They will likely cause issues (like this one) at some point when using AssertJ in an OSGi environment, because the bundles trying to use such a class won't be able to load it.

**Comment by kriegfrj** [comment]
It would be helpful if Bnd were a little more forthcoming about which classes in those packages are generating the private references.

**Comment by joel-costigliola** [comment]
I'm not sure if that's a silly suggestion but I'm not against exporting more classes (if not all) in the manifest if that can solve thes never ending OSGi issues. I would not suggest that if AssertJ was a library use in production code but it is meant for testing only so even if the export is loose it won't impact production artifacts.

**Comment by bjhargrave** [comment]
So, I went back to my composite class loader attempt for another try. Turns out I got confused by a test case in assertj which failed when testing my fix. The issue was that the test class, SoftAssertionsMultipleProjectsTest, was only using package private access for the assert classes being proxied and those classes super types and interfaces. ByteBuddy was in fact defining the generated proxy class in the composite class loader but the generated proxy class' super type was only package private in the test package, so the VM failed to load the class with an IllegalAccessError. When I changed the test assert class and the types it used to all be public, everything worked correctly. So I have made PR #1968 to added the composite class loader which is only used for assert classes which come from a different class loader than the AssertJ classes. I tested this change out with the osgi-test custom assertion in the opening comment and it works great. So I think the PR will solve a number of soft proxy issues when defining custom assertions in a different class loader than the AssertJ classes.
